listenbrainz_spark/data/dump.py

```diff
@@ -6,37 +7,80 @@
 
 import listenbrainz_spark.config as config
 
+from datetime import datetime
 from hdfs.util import HdfsError
-from listenbrainz_spark import spark, sc
+from listenbrainz_spark import spark, sc, hdfs_connection
+from listenbrainz_spark.constants import LAST_FM_FOUNDING_YEAR
+from listenbrainz_spark.schema import convert_listen_to_row, listen_schema
+from pyspark.sql.functions import year
 
 FORCE = True
 
-def copy_to_hdfs(archive):
-    pxz_command = ['pxz', '--decompress', '--stdout', archive, '-T4']
+def _is_listens_file(filename):
+    """ Check if passed filename is a file which contains listens
+
+    Args:
+        filename (str): the name of the file
+
+    Returns:
+        bool: True if the file contains listens, False otherwise
+    """
+    return filename.endswith('.listens')
+
+
+def copy_to_hdfs(archive, threads=4):
+    """ Create Spark Dataframes from a listens dump and save it to HDFS.
+
+    Args:
+        archive (str): the path to the listens dump
+        threads (int): the number of threads to use for decompression of the archive
+    """
+    pxz_command = ['pxz', '--decompress', '--stdout', archive, '-T{}'.format(threads)]
     pxz = subprocess.Popen(pxz_command, stdout=subprocess.PIPE)
-    hdfs_client = hdfs.InsecureClient(config.HDFS_NAMENODE_URI, user='root')
     destination_path = os.path.join('/', 'data', 'listenbrainz')
     if FORCE:
         print('Removing data directory if present...')
-        hdfs_client.delete(destination_path, recursive=True)
+        hdfs_connection.client.delete(destination_path, recursive=True)
         print('Done!')
 
-    print('Creating remote directory...')
-    hdfs_client.makedirs(destination_path)
-    print('Done!')
-
-    print('Uploading...')
+    print('Creating Listens RDD...')
+    listens_rdd = sc.parallelize([])
     with tarfile.open(fileobj=pxz.stdout, mode='r|') as tar:
         for member in tar:
-            if member.isfile():
+            if member.isfile() and _is_listens_file(member.name):
+                print('Loading %s...' % member.name)
                 tar.extract(member)
-                print("Uploading %s to %s..." % (os.path.join(destination_path, member.name), member.name))
-                hdfs_client.upload(os.path.join(destination_path, member.name), member.name)
+                listens_rdd = listens_rdd.union(sc.textFile(member.name).map(json.loads))
+                listens_rdd.cache().count()
+                print("Done!")
                 os.remove(member.name)
     print("Done!")
 
+    print("Creating Listens Dataframe...")
```

Comment:
Hmmm. Creating a data from of all the listens and then creating yearly/monthly frames -- am I reading that right? Is there any way for us to skip this very time/resource intensive step and go directly to monthly frames? Or is the all listens frame needed for something else later?